### ISSUE TITLE
piping to gulp-uglify

### ISSUE THREAD

**Issue by kasperlewau** [ragingwind/gulp-crisper#15]
Hiya! 

So, I was thinking I would take a stab at rewiring a Grunt task into a Gulp task. 

One obstacle I've now come across is when I want to run the following chain of events: 

```
vulcanize
└── {DIST}
└── crisper
    └── uglify
        └── {DIST}
```

``` js
import gulp from 'gulp';
import es from 'event-stream';

import source from 'vinyl-source-stream';
import buffer from 'vinyl-buffer';

import vulcanize from 'gulp-vulcanize';
import crisper from 'gulp-crisper';
import uglify from 'gulp-uglify';

import { wcEntry, uglifyConf } from '../config';

/**
 * Run the file at {wcEntry} through Vulcanize, Crisper & Uglify
 * to produce a sweet-as-pie WebComponents build in ./dist/static
 */
gulp.task('wc', () => {
    /**
     * Vulcanize the entry file and write html to ./dist/static.
     */
    const vulcanized = gulp.src(wcEntry)
        .pipe(vulcanize({
            stripExcludes: './http-pub/bower/polymer/polymer.html',
            inlineScripts: true,
            inlineCss: true,
            stripComments: true
        }))
        .pipe(gulp.dest('./dist/static'));

    /**
     * Crisp the vulcanized output and write minified js to ./dist/static.
     */
    const crisped = vulcanized
        .pipe(crisper({
            scriptInHead: false
        }))
        .pipe(uglify())
        .on('error', console.log)
        .pipe(gulp.dest('./dist/static'))

    return es.merge(vulcanized, crisped);
});
```

I'm not sure whether it has to do with uglify being greedy, or crisper being pushy (that's about as far as my current understanding of Streams go) - but I'm getting the full `vulcanize` output into `uglify`. 

It's as if `crisper` hasn't yet had a chance at doing it's thing and giving me `.js`. 

Example error: 

```
{ [Error: /Users/kasperlewau/code/foo/html/dist/static/shared.imports.html: SyntaxError: Unexpected token: operator (<)]
  message: '/Users/kasperlewau/code/foo/html/dist/static/shared.imports.html: SyntaxError: Unexpected token: operator (<)',
  fileName: '/Users/kasperlewau/code/foo/html/dist/static/shared.imports.html',
  lineNumber: 1,
  stack: 'Error\n    at new JS_Parse_Error (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:1534:18)\n    at js_error (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:1542:11)\n    at croak (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2089:9)\n    at token_error (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2097:9)\n    at unexpected (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2103:9)\n    at expr_atom (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2630:9)\n    at maybe_unary (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2792:19)\n    at expr_ops (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2827:24)\n    at maybe_conditional (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2832:20)\n    at maybe_assign (eval at <anonymous> (/Users/kasperlewau/code/foo/html/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2856:20)',
  showStack: false,
  showProperties: true,
  plugin: 'gulp-uglify' }
```

The operator (`<`) in question comes from line 1 of the `vulcanize` output: 

``` html
<html><head><!-- polymer project copyright notice -->
```

---

If I switch out [these lines](https://github.com/ragingwind/gulp-crisper/blob/master/index.js) in `crisper/index.js` for this: 

``` js
// Object.keys(split).forEach(function (type) {
//  if (split[type]) {
//    stream.push(splitFile(file, splitfile[type], split[type]));
//  }
// });

stream.push(splitFile(file, splitfile.js, split.js));
```

---
#### It works just fine™!

...because I'm now only outputting JS to `uglify`. Obviously without the `html` part, you'd effectively make having two streams required to get all the output you need, so obviously that's nota

**tl;dr;** Would you consider an option to `crisper` to _only_ output either `js` or `html` so that we can plug the individual parts into further optimizer(s) for each given language? 

I'd be happy to get a PR going for that.

**Comment by ragingwind** [comment]
As far as I known that vulcanize strip out of them like uglify or if you would like to uglify those of js files, I recommend to use gulp-if

**Comment by kasperlewau** [comment]
Myeah, come to think of it I'd be better off looking at a way to _pick_ the individual files that `crisper` outputs and process them further depending on some sort of conditional statement. 

That would make the `crisper` stream reusable, as opposed to having two crisper passes (one for js, one for html). 

Something like this (tried this out, didn't get it to fly though): 

``` js
let jsTime = false;
let htmlTime = false;

gulp.src('foo.html')
  .pipe(crisper())
  .pipe(es.map(function (data, cb) {
    if (/\.js/.test(data.history[0])) {
      jsTime = true;
    } else {
      htmlTime = true;
    }

    cb(null, data);
  })
  .pipe(gulpif(jsTime, uglify({}))
  .pipe(gulpif(htmlTime, htmlmin({}))
  .pipe(gulp.dest('./dist'));
```

Regardless, I reckon we can close the issue. I just need to get comfy with streams in general and figure out my approach. Cheers!

**Comment by ragingwind** [comment]
thanks for the idea

**Comment by kasperlewau** [comment]
for anyone else concerned with this, I think this is the slimmest I can get it (right now) and it works. 👌 

``` js
const htmlPath = lazypipe()
  .pipe(htmlmin)
  .pipe(gulp.dest, './');

const jsPath = lazypipe()
  .pipe(uglify)
  .pipe(gulp.dest, './');

gulp.src('foo.html')
  .pipe(crisper)
  .pipe(gulpif(/\.html/, htmlPath(), jsPath());
```

**Comment by mikegoodstadt** [comment]
@kasperlewau - I think there is a missing end bracket:
`.pipe(gulpif( /\.html/, htmlPath(), jsPath() ));`
Thanks for the guidance :D !
